src/lib/pickles/test/test_wrap.ml

```diff
@@ -0,0 +1,334 @@
+          Sponge.absorb sponge
+            (`Field (Impl.Field.constant sponge_digest_before_evaluations)) ;
+          sponge
+        in
+
+        (* Call finalisation with all of the required details *)
+        Step_verifier.finalize_other_proof
+          (module Nat.N0)
+          ~step_domains:
+            (`Known [ { h = Pow_2_roots_of_unity vk.domain.log_size_of_group } ])
+          ~zk_rows:3 ~sponge ~prev_challenges:[] deferred_values evals
+      in
+
+      (* Read the boolean result from the circuit and make it available
+         to the OCaml world. *)
+      Impls.Step.(As_prover.(fun () -> read Boolean.typ res)) )
+  |> Or_error.ok_exn
+
+(* Common srs value for all tests *)
+let srs =
+  Kimchi_bindings.Protocol.SRS.Fp.create (1 lsl Common.Max_degree.step_log2)
+
+type example =
+     Kimchi_bindings.Protocol.SRS.Fp.t
+  -> Kimchi_bindings.Protocol.Index.Fp.t
+     * Pasta_bindings.Fp.t list
+     * ( Pasta_bindings.Fq.t Kimchi_types.or_infinity
+       , Pasta_bindings.Fp.t )
+       Kimchi_types.proof_with_public
+
+module type SETUP = sig
+  val example : example
+
+  (* Feature flags tused for backend proof *)
+  val actual_feature_flags : bool Plonk_types.Features.t
+end
+
+(* [Make] is the test functor.
+
+   Given a test setup, compute different test configurations and define 3
+   test for said configurations. *)
+module Make (S : SETUP) = struct
+  (* Generate foreign field multiplication test backend proof using Kimchi,
+     obtaining the proof and corresponding prover index.
+
+     Note: we only want to pay the cost of generating this proof once and
+     then reuse it many times for the different recursive proof tests. *)
+  let index, public_input, proof = S.example srs
+
+  (* Obtain verifier key from prover index and convert backend proof to
+     snarky proof *)
+  let vk = Kimchi_bindings.Protocol.VerifierIndex.Fp.create index
+
+  let proof = Backend.Tick.Proof.of_backend_with_public_evals proof
+
+  let test_feature_flags_configs =
+    generate_test_feature_flag_configs S.actual_feature_flags
+
+  let runtest feature_flags =
+    run_recursive_proof_test S.actual_feature_flags feature_flags public_input
+      vk proof
+
+  let test_true_is_yes () =
+    assert (runtest test_feature_flags_configs.true_is_yes)
+
+  let test_true_is_maybe () =
+    assert (runtest test_feature_flags_configs.true_is_maybe)
+
+  let test_all_maybes () = assert (runtest test_feature_flags_configs.all_maybes)
+
+  let tests =
+    let open Alcotest in
+    [ test_case "true -> yes" `Quick test_true_is_yes
+    ; test_case "true -> maybe" `Quick test_true_is_maybe
+    ; test_case "all maybes" `Quick test_all_maybes
+    ]
+end
+
+(* Small combinators to lift gate example signatures to the expected
+   signatures for the tests. This amounts to generating the list of public
+   inputs from either no public inputs, a single one or a pair of inputs
+   returned by the gate example. *)
+
+let no_public_input gate_example srs =
```

Comment:
Can these three combinators please be explicitly typed? Also perhaps `pack_public_input`/`sel_public_input` or something like that would make it more clear that these are utilities... I feel that overusing generic names in the codebase is making searching through code more tedious, although in this particular case it's clear it's a utility of course.

Reply:
I'm not sure what you actually would like here. Do you want to write that this function has type `('a -> 'b * 'c) -> 'a -> 'b * 'd list * 'c`?  The usual practice in `OCaml` is not to write types for local functions unless necessary,  and, personally, I don't see the value in this particular case.

Regarding naming, I don't see why the names you give makes it clearer that these are utilities. Your remark related to the codebase might be relevant nonetheless. I'll be happy to discuss this issue with you.

Reply:
>The usual practice in OCaml is not to write types for local functions unless necessary, and, personally, I don't see the value in this particular case.

I'm fine to resolve this conversation if it's general style element, perhaps there's too much subjectivity in my approach.

Explaining what I meant though, I expect the function name + type to tell me 90% about what it does. `('a -> 'b * 'c) -> 'a -> 'b * 'd list * 'c` I like a lot because it's very polymorphic and I immediately recognise it as some kind of constructor. While the name `public_input_1` to me suggests it's choosing the first element of a tuple from some `PublicInput` type... But then again ocaml conventions in the repo rely on (1) implicit imports/reimports as opposed to explicit ones, (2) highly duplicated naming, (3) duplicated type definitions, (4) implicit side effects. So maybe my strategy works better for rust/haskell, and reading ocaml needs a different approach...

Reply:
Would you like names `one_public_input` and `two_public_inputs` better?

Reply:
I changed names.